java/ql/src/semmle/code/java/StringFormat.qll

```diff
@@ -425,6 +440,16 @@ private class PrintfFormatString extends FormatString {
     result > count(int i | fmtSpecRefersToSequentialIndex(i)) and
     not result = fmtSpecRefersToSpecificIndex(_)
   }
+
+  override int getAnArgUsageOffset(int argNo) {
+    argNo = fmtSpecRefersToSpecificIndex(result)
+    or
+    fmtSpecRefersToSequentialIndex(result) and
+    argNo = count(int i | i < result and fmtSpecRefersToSequentialIndex(i))
+    or
+    fmtSpecRefersToPrevious(result) and
+    argNo = count(int i | i < result and fmtSpecRefersToSequentialIndex(i)) - 1
+  }
```

Comment:
The "refers-to-previous" disjunct looks incorrect to me. It looks like you assume that the `%<` always refers to a sequential index, but it can also refer to another `%<` or a numbered format specifier. Also, this implementation is needlessly quadratic, which is slightly jarring to me (I know that the number of format specifiers in format strings likely aren't big, but still).

Reply:
See https://docs.oracle.com/javase/7/docs/api/java/util/Formatter.html